Authentication does not work with Drupal

In Nuxt 2 & Axios, there was an option (enabled by default), [proxyHeaders](https://axios.nuxtjs.org/options/#proxyheaders), which would take care of cookie based auth.

In Nuxt 3, there is a Nuxt composable, [useRequestHeaders](https://nuxt.com/docs/api/composables/use-request-headers#userequestheaders), which can be used with this module, like this:
`await fetchPage(useRoute().path, { query: useRoute().query, credentials: 'include', headers: useRequestHeaders(['cookie']) })`

We want to simplify cookie auth, to work out-of-the-box, so **we'll enable it by default** for every request to Drupal, **but it should be overridable** by specifying `headers` in an object in 2nd param of fetchPage and fetchMenu (`useFetchOptions`).

Turns out "npm run prepack" was failing on some type-error:

          src/module.ts(39,7): error TS2322: Type 'UseFetchOptions<any, _Transform<any, any>, KeyOfRes<_Transform<any, any>>, string & {}, "get" | "head" | "patch" | "post" | "put" | "delete" | "connect" | "options" | "trace">' is not assignable to type '{ credentials: string; }'.
            Types of property 'credentials' are incompatible.
              Type 'RequestCredentials | Ref<RequestCredentials | undefined> | undefined' is not assignable to type 'string'.

I did not get it working with that default. :-( So I dropped it for now, possibly, this is why it's not working:
https://github.com/drunomics/nuxtjs-drupal-ce/pull/127/commits/280b5c73c94214112b38ed1c16cd2f21bc63fba1

--- a/playground/nuxt.config.ts
+++ b/playground/nuxt.config.ts
@@ -7,6 +7,5 @@ export default defineNuxtConfig({
   ],
   drupalCe: {
     baseURL: process.env.DRUPAL_BASE_URL || 'https://8080-drunomics-lupusdecouple-fypxwm6rhgb.ws-eu89.gitpod.io/ce-api'
-  },
-  ssr: false
+  }
 })

--- a/playground/nuxt.config.ts
+++ b/playground/nuxt.config.ts
@@ -6,6 +6,7 @@ export default defineNuxtConfig({
     DrupalCe
   ],
   drupalCe: {
-    baseURL: process.env.DRUPAL_BASE_URL || 'https://8080-drunomics-lupusdecouple-xxxxxxxxxx.ws-euxx.gitpod.io/ce-api'
-  }
+    baseURL: process.env.DRUPAL_BASE_URL || 'https://8080-drunomics-lupusdecouple-fypxwm6rhgb.ws-eu89.gitpod.io/ce-api'
+  },
+  ssr: false
 })

--- a/src/module.ts
+++ b/src/module.ts
@@ -20,9 +20,7 @@ export default defineNuxtModule<ModuleOptions>({
     baseURL: 'https://8080-shaal-drupalpod-xxxxxxxxxxx.ws-xxxx.gitpod.io/ce-api',
     menuEndpoint: 'api/menu_items/$$$NAME$$$',
     customErrorPages: false,
-    fetchOptions: {
-      credentials: 'include'
-    },
+    fetchOptions: {},
     fetchProxyHeaders: ['cookie']
   },
   setup (options, nuxt) {

--- a/src/module.ts
+++ b/src/module.ts
@@ -1,11 +1,14 @@
 import { fileURLToPath } from 'url'
 import { defineNuxtModule, addPlugin, createResolver, addImportsDir } from '@nuxt/kit'
+import { UseFetchOptions } from 'nuxt/dist/app/composables'
 
 export interface ModuleOptions {
   baseURL: string,
   menuEndpoint: string,
   addRequestContentFormat?: string,
   customErrorPages: boolean,
+  fetchOptions: UseFetchOptions<any>,
+  fetchProxyHeaders: string[],
 }
 
 export default defineNuxtModule<ModuleOptions>({
@@ -16,7 +19,11 @@ export default defineNuxtModule<ModuleOptions>({
   defaults: {
     baseURL: 'https://8080-shaal-drupalpod-xxxxxxxxxxx.ws-xxxx.gitpod.io/ce-api',
     menuEndpoint: 'api/menu_items/$$$NAME$$$',
-    customErrorPages: false
+    customErrorPages: false,
+    fetchOptions: {
+      credentials: 'include'
+    },
+    fetchProxyHeaders: ['cookie']
   },
   setup (options, nuxt) {
     const { resolve } = createResolver(import.meta.url)
@@ -28,7 +35,9 @@ export default defineNuxtModule<ModuleOptions>({
       baseURL: options.baseURL,
       menuEndpoint: options.menuEndpoint,
       addRequestContentFormat: options.addRequestContentFormat,
-      customErrorPages: options.customErrorPages
+      customErrorPages: options.customErrorPages,
+      fetchOptions: options.fetchOptions,
+      fetchProxyHeaders: options.fetchProxyHeaders
     }
   }
 })

--- a/src/runtime/composables/useDrupalCe.ts
+++ b/src/runtime/composables/useDrupalCe.ts
@@ -1,29 +1,49 @@
 import { callWithNuxt } from '#app'
+import { defu } from 'defu'
 import { useRuntimeConfig, useState, useFetch, navigateTo, createError, useRoute, h, resolveComponent, setResponseStatus, useNuxtApp } from '#imports'
+import { UseFetchOptions } from 'nuxt/dist/app/composables'
+
 export const useDrupalCe = () => {
+
+  const config = useRuntimeConfig().public.drupalCe
+
+  /**
+   * Processes the given fetchOptions to apply module defaults
+   * @param fetchOptions Optional Nuxt useFetch options
+   * @returns UseFetchOptions<any>
+   */
+  const processFetchOptions = (fetchOptions:UseFetchOptions<any> = {}) => {
+    fetchOptions.baseURL = fetchOptions.baseURL ?? config.baseURL
+    fetchOptions = defu(fetchOptions, config.fetchOptions)
+
+    // Apply the request headers of current request, if configured.
+    if (config.fetchProxyHeaders) {
+      fetchOptions.headers = defu(fetchOptions.headers ?? {}, useRequestHeaders(config.fetchProxyHeaders))
+    }
+    return fetchOptions
+  }
+
   /**
    * Fetches page data from Drupal, handles redirects, errors and messages
    * @param path Path of the Drupal page to fetch
    * @param useFetchOptions Optional Nuxt useFetch options
    */
-  const fetchPage = async (path: string, useFetchOptions = {}) => {
+  const fetchPage = async (path: string, useFetchOptions:UseFetchOptions<any> = {}) => {
     const nuxtApp = useNuxtApp()
-    const config = useRuntimeConfig()
-    const baseURL = config.public.drupalCe.baseURL
 
     // Workaround for issue - useState is not available after async call (Nuxt instance unavailable)
     const pageState = useState(`page-${path}`, () => {})
-
-    useFetchOptions.query = useFetchOptions.query ?? {}
     useFetchOptions.key = `page-${path}`
-    useFetchOptions.baseURL = baseURL
+    useFetchOptions = processFetchOptions(useFetchOptions)
 
-    if (config.public.drupalCe.addRequestContentFormat) {
-      useFetchOptions.query.addRequestContentFormat = config.public.drupalCe.addRequestContentFormat
+    if (config.addRequestContentFormat) {
+      useFetchOptions.query = useFetchOptions.query ?? {}
+      useFetchOptions.query.addRequestContentFormat = config.addRequestContentFormat
     }
 
+    console.log(useFetchOptions)
     const { data: page, error } = await useFetch(path, useFetchOptions)
-
+    
     if (page?.value?.redirect) {
       await navigateTo(page.value.redirect.url, {
         external: page.value.redirect.external,
@@ -32,7 +52,7 @@ export const useDrupalCe = () => {
       return
     }
 
-    if (error.value && (!error.value?.data?.content || config.public.drupalCe.customErrorPages)) {
+    if (error.value && (!error.value?.data?.content || config.customErrorPages)) {
       throw createError({ statusCode: error.value.status, statusMessage: error.value.message, data: error.value.data, fatal: true })
     }
 
@@ -50,23 +70,19 @@ export const useDrupalCe = () => {
   /**
    * Fetches menu data from Drupal (configured by menuEndpoint option), handles errors
    * @param name Menu name being fetched
+   * @param useFetchOptions Optional Nuxt useFetch options
    */
-  const fetchMenu = async (name: string) => {
-    const config = useRuntimeConfig()
-    const baseURL = config.public.drupalCe.baseURL
-    const menuEndpoint = config.public.drupalCe.menuEndpoint
-    const menuPath = menuEndpoint.replace('$$$NAME$$$', name)
+  const fetchMenu = async (name: string, useFetchOptions:UseFetchOptions<any> = {}) => {
+    const menuPath = config.menuEndpoint.replace('$$$NAME$$$', name)
+    useFetchOptions = processFetchOptions(useFetchOptions)
+    useFetchOptions.key = `menu-${name}`
 
-    const { data: menu, error } = await useFetch(menuPath, {
-      key: `menu-${name}`,
-      baseURL
-    })
+    const { data: menu, error } = await useFetch(menuPath, useFetchOptions)
 
     if (error.value) {
       errorMenuHandler(error)
       return
     }
-
     return menu
   }
 

--- a/src/runtime/composables/useDrupalCe.ts
+++ b/src/runtime/composables/useDrupalCe.ts
@@ -41,7 +41,6 @@ export const useDrupalCe = () => {
       useFetchOptions.query.addRequestContentFormat = config.addRequestContentFormat
     }
 
-    console.log(useFetchOptions)
     const { data: page, error } = await useFetch(path, useFetchOptions)
     
     if (page?.value?.redirect) {

